gossipd/gossipd.c

```diff
@@ -1684,39 +1702,46 @@ static void dump_gossip(struct peer *peer)
 	maybe_create_next_scid_reply(peer);
 }
 
+/*~ Our timer callbacks take a single argument, so we marshall everything
+ * we need into this structure: */
+struct local_cupdate {
+	struct daemon *daemon;
+	struct local_chan *local_chan;
+
+	bool disable;
+	bool even_if_identical;
+
+	u16 cltv_expiry_delta;
+	struct amount_msat htlc_minimum, htlc_maximum;
+	u32 fee_base_msat, fee_proportional_millionths;
+};
+
 /*~ This generates a `channel_update` message for one of our channels.  We do
  * this here, rather than in `channeld` because we (may) need to do it
- * ourselves anyway if channeld dies, or when we refresh it once a week. */
-static void update_local_channel(struct daemon *daemon,
-				 const struct chan *chan,
-				 int direction,
-				 bool disable,
-				 u16 cltv_expiry_delta,
-				 struct amount_msat htlc_minimum,
-				 u32 fee_base_msat,
-				 u32 fee_proportional_millionths,
-				 struct amount_msat htlc_maximum,
-				 const char *caller)
+ * ourselves anyway if channeld dies, or when we refresh it once a week,
+ * and so we can avoid creating redundant ones. */
+static void update_local_channel(struct local_cupdate *lc /* frees! */)
 {
+	struct daemon *daemon = lc->daemon;
 	secp256k1_ecdsa_signature dummy_sig;
 	u8 *update, *msg;
-	u32 timestamp = gossip_time_now(daemon->rstate).ts.tv_sec;
+	u32 timestamp = gossip_time_now(daemon->rstate).ts.tv_sec, next;
 	u8 message_flags, channel_flags;
+	struct chan *chan = lc->local_chan->chan;
+	struct half_chan *hc;
+	const int direction = lc->local_chan->direction;
+
+	/* Discard existing timer. */
+	lc->local_chan->channel_update_timer
+		= tal_free(lc->local_chan->channel_update_timer);
 
 	/* So valgrind doesn't complain */
 	memset(&dummy_sig, 0, sizeof(dummy_sig));
 
-	/* BOLT #7:
-	 *
-	 * The origin node:
-	 *...
-	 *   - MUST set `timestamp` to greater than 0, AND to greater than any
-	 *     previously-sent `channel_update` for this `short_channel_id`.
-	 *     - SHOULD base `timestamp` on a UNIX timestamp.
-	 */
-	if (is_halfchan_defined(&chan->half[direction])
-	    && timestamp == chan->half[direction].bcast.timestamp)
-		timestamp++;
+	/* Create an unsigned channel_update: we backdate enables, so
+	 * we can always send a disable in an emergency. */
+	if (!lc->disable)
+		timestamp -= daemon->gossip_min_interval;
```

Comment:
This will hit the "back off and retry" logic:

```
		/* Is it too soon to send another update? */
		next = hc->bcast.timestamp
			+ GOSSIP_MIN_INTERVAL(daemon->rstate->dev_fast_gossip);
```